### PR TITLE
Improve mouse-based load inhibition at player loading screen

### PR DESCRIPTION
**Pull request by UselessToucan**
We probably do not want to move from PlayerLoader to Player if the user pressed mouse button down on one of the sliderbars and did not release it yet.

I can't provide the video right now, but here are the steps to reproduce:
1. Perform MouseDown on one of the VisualSettings overlay sliderbars.
2. Move the cursor outside of the overlay while holding mouse button pressed.

### CODE REVIEW

**Review by peppy — CHANGES_REQUESTED** [review]
Still not a huge fan of exposing IsMouseDown.

**Comment by smoogipoo** [comment]
How about just using `Drawable.IsDragged` on the sliderbars?

**Comment by UselessToucan** [comment]
@smoogipoo I tried to use `Drawable.IsDragged`. It does not solve the problem since it becomes false earlier than I release the mouse button.

**Comment by peppy** [comment]
According to its implementation, it shouldn't.

**Comment by peppy** [comment]
Two potential better solutions:

In `PlayerLoader.cs`

```csharp
var mouseDownPos = GetContainingInputManager().CurrentState.Mouse.NativeState.PositionMouseDown;
if (player.LoadState != LoadState.Ready || visualSettings.IsHovered || (mouseDownPos != null && visualSettings.ReceiveMouseInputAt(mouseDownPos.Value)))
{
    Schedule(pushWhenLoaded);
    return;
}
```

In `VisualSettings.cs`
```csharp
private bool mouseDown;

public override bool ReceiveMouseInputAt(Vector2 screenSpacePos) => mouseDown || base.ReceiveMouseInputAt(screenSpacePos);

protected override bool OnMouseDown(InputState state, MouseDownEventArgs args)
{
    mouseDown = true;
    return base.OnMouseDown(state, args);
}

protected override bool OnMouseUp(InputState state, MouseUpEventArgs args)
{
    mouseDown = false;
    return base.OnMouseUp(state, args);
}
```

Both result in very localised implementations, which handle more than just sliders at the same time. Not sure which I prefer.

**Comment by UselessToucan** [comment]
I prefer the first one. In my opinion, it is easier.

**Comment by peppy** [comment]
@smoogipoo would like a second opinion on this. it's likely going to pop up in more places going forward if anything (volume overlay being one such example).

**Comment by smoogipoo** [comment]
At the very least I believe these should be inverted. I.e. instead of `visualSettings.ReceiveMouseInputAt(...)` it should use `PlayerLoader.ReceiveMouseInputAt`. Same thing with `IsHovered`.

Furthermore, why does `IsDragged` not work? Sounds like an issue...

**Comment by peppy** [comment]
I checked on `IsDragged`: you can potentially leave the visual settings area before the drag lenience has been hit, at which point there's no dragged state activated.

Inverting could have unwanted side effects, like hovering the toolbar or chat or something. Not sure if we want that.

**Comment by smoogipoo** [comment]
The idea being that those places should not handle OnHover if they don't want player's loading to stop.

I mean there really isn't any other way - either you have the player check for each component that it needs, or you have some "hey I'm currently handling input" state be passed down to the `PlayerLoader`.

I believe the chat is an overlaycontainer, yes? In which case it probably already does block passthrough hover and I think that makes more sense than pushing a player while chat is open?

**Comment by peppy** [comment]
It's definitely worth a try to see how it feels.

**Comment by peppy** [comment]
Reworked to add debounce and work in a more general fashion. @smoogipoo requesting review as this is now mostly my code ;x

**Review by peppy — DISMISSED** [review]

File: osu.Game/Screens/Play/PlayerLoader.cs [hunk 1]
```diff
@@ -116,29 +118,52 @@ protected override void LogoArriving(OsuLogo logo, bool resuming)
...
+        }
+
+        private ScheduledDelegate pushDebounce;
+
+        private bool readyForPush => player.LoadState == LoadState.Ready && IsHovered && (!GetContainingInputManager().CurrentState.Mouse.HasAnyButtonPressed || weHandledMouseDown);
+
         private void pushWhenLoaded()
         {
-            if (player.LoadState != LoadState.Ready || visualSettings.IsHovered)
+            Schedule(pushWhenLoaded);
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/PlayerLoader.cs]: This scheduling is still going on during gameplay

File: osu.Game/Screens/Play/PlayerLoader.cs [hunk 2]
```diff
@@ -116,29 +118,52 @@ protected override void LogoArriving(OsuLogo logo, bool resuming)
...
-                Schedule(pushWhenLoaded);
+                pushDebounce?.Cancel();
+                pushDebounce = null;
                 return;
             }
 
-            contentOut();
-
-            this.Delay(250).Schedule(() =>
+            if (pushDebounce == null) pushDebounce = Scheduler.AddDelayed(() =>
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/PlayerLoader.cs]: Put this on a separate line please.

**Review by smoogipoo — DISMISSED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**